DLQ - PSQLException: ERROR: value too long for type character varying(255)

### Basic information

* Axon Framework version: 4.6.3
* JDK version:  11.0.18

```
2023-06-14 08:19:10.065  INFO [query,,] 7 --- [ses-internal]-3] o.a.e.d.jpa.JpaSequencedDeadLetterQueue  : Storing DeadLetter (id: [9835881b-8c83-4973-8a9b-73f5e7fe71c3]) for sequence [21-102-464-1] with index [0] in processing group [cases-internal].
2023-06-14 08:19:10.090  WARN [query,,] 7 --- [ses-internal]-3] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 22001
2023-06-14 08:19:10.107  WARN [query,,] 7 --- [ses-internal]-3] o.a.eventhandling.pooled.WorkPackage     : Error while processing batch in Work Package [14]-[cases-internal]. Aborting Work Package...
2023-06-14 08:19:12.942  INFO [query,,] 7 --- [ses-internal]-0] o.a.eventhandling.pooled.Coordinator     : Processor [cases-internal] claimed 1 new segments for processing
org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; nested exception is org.hibernate.exception.DataException: could not execute statement
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:280) ~[spring-orm-5.3.27.jar:5.3.27]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:233) ~[spring-orm-5.3.27.jar:5.3.27]
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:566) ~[spring-orm-5.3.27.jar:5.3.27]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:743) ~[spring-tx-5.3.27.jar:5.3.27]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:711) ~[spring-tx-5.3.27.jar:5.3.27]
	at org.axonframework.spring.messaging.unitofwork.SpringTransactionManager.commitTransaction(SpringTransactionManager.java:86) ~[axon-spring-4.6.3.jar:4.6.3]
	at org.axonframework.spring.messaging.unitofwork.SpringTransactionManager$1.commit(SpringTransactionManager.java:69) ~[axon-spring-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.UnitOfWork.lambda$attachTransaction$0(UnitOfWork.java:274) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.MessageProcessingContext.notifyHandlers(MessageProcessingContext.java:72) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.BatchingUnitOfWork.lambda$notifyHandlers$2(BatchingUnitOfWork.java:161) ~[axon-messaging-4.6.3.jar:4.6.3]
	at java.base/java.util.Iterator.forEachRemaining(Unknown Source) ~[na:na]
	at org.axonframework.messaging.unitofwork.BatchingUnitOfWork.notifyHandlers(BatchingUnitOfWork.java:161) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.AbstractUnitOfWork.changePhase(AbstractUnitOfWork.java:236) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.AbstractUnitOfWork.commitAsRoot(AbstractUnitOfWork.java:87) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.AbstractUnitOfWork.commit(AbstractUnitOfWork.java:75) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.messaging.unitofwork.BatchingUnitOfWork.executeWithResult(BatchingUnitOfWork.java:117) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.eventhandling.AbstractEventProcessor.processInUnitOfWork(AbstractEventProcessor.java:166) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.eventhandling.pooled.WorkPackage.processEvents(WorkPackage.java:305) ~[axon-messaging-4.6.3.jar:4.6.3]
	at org.axonframework.eventhandling.pooled.WorkPackage.lambda$scheduleWorker$5(WorkPackage.java:270) ~[axon-messaging-4.6.3.jar:4.6.3]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
	at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
Caused by: org.hibernate.exception.DataException: could not execute statement
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:115) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3937) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at java.base/java.util.LinkedHashMap.forEach(Unknown Source) ~[na:na]
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1407) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:489) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3303) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2438) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:449) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:183) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:40) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:281) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:562) ~[spring-orm-5.3.27.jar:5.3.27]
	... 22 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: value too long for type character varying(255)
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:167) ~[postgresql-42.3.8.jar:42.3.8]
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135) ~[postgresql-42.3.8.jar:42.3.8]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61) ~[HikariCP-4.0.3.jar:na]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java) ~[HikariCP-4.0.3.jar:na]
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197) ~[hibernate-core-5.6.15.Final.jar:5.6.15.Final]
	... 43 common frames omitted
2023-06-14 08:19:10.092  INFO [query,,] 7 --- [ses-internal]-3] o.h.e.j.b.internal.AbstractBatchImpl     : HHH000010: On release of batch it still contained JDBC statements
2023-06-14 08:19:10.090 ERROR [query,,] 7 --- [ses-internal]-3] o.h.engine.jdbc.spi.SqlExceptionHelper   : ERROR: value too long for type character varying(255)
```

Using the following DDL:
```
create table dead_letter_entry
(
  dead_letter_id       varchar(255) not null
    primary key,
  cause_message        varchar(255),
  cause_type           varchar(255),
  diagnostics          bytea,
  enqueued_at          timestamp    not null,
  last_touched         timestamp,
  aggregate_identifier varchar(255),
  event_identifier     varchar(255) not null,
  message_type         varchar(255) not null,
  meta_data            bytea,
  payload              bytea        not null,
  payload_revision     varchar(255),
  payload_type         varchar(255) not null,
  sequence_number      bigint,
  time_stamp           varchar(255) not null,
  token                bytea,
  token_type           varchar(255),
  type                 varchar(255),
  processing_group     varchar(255) not null,
  processing_started   timestamp,
  sequence_identifier  varchar(255) not null,
  sequence_index       bigint       not null
);

create index dead_letter_entry_processing_group_idx
  on dead_letter_entry (processing_group);

create index dead_letter_entry_processing_group_seq_identifier_idx
  on dead_letter_entry (processing_group, sequence_identifier);

create unique index dead_letter_entry_processing_group_seq_identifier_seq_index_idx
  on dead_letter_entry (processing_group, sequence_identifier, sequence_index);
```

I assume the error message is too long. Is there a missing error message truncation, or would it be more appropriate to remove the size limit of the varchar?

--- a/config/src/main/java/org/axonframework/config/EventProcessingModule.java
+++ b/config/src/main/java/org/axonframework/config/EventProcessingModule.java
@@ -49,6 +49,7 @@ import org.axonframework.messaging.deadletter.Decisions;
 import org.axonframework.messaging.deadletter.EnqueuePolicy;
 import org.axonframework.messaging.deadletter.SequencedDeadLetterProcessor;
 import org.axonframework.messaging.deadletter.SequencedDeadLetterQueue;
+import org.axonframework.messaging.deadletter.ThrowableCause;
 import org.axonframework.messaging.interceptors.CorrelationDataInterceptor;
 import org.axonframework.messaging.unitofwork.RollbackConfiguration;
 import org.axonframework.messaging.unitofwork.RollbackConfigurationType;
@@ -170,9 +171,10 @@ public class EventProcessingModule
     );
     @SuppressWarnings({"unchecked", "rawtypes"})
     private final Component<EnqueuePolicy<EventMessage<?>>> defaultDeadLetterPolicy = new Component<>(
-            () -> configuration,
-            "deadLetterPolicy",
-            c -> c.getComponent(EnqueuePolicy.class, () -> (letter, cause) -> Decisions.enqueue(cause))
+            () -> configuration, "deadLetterPolicy",
+            c -> c.getComponent(EnqueuePolicy.class,
+                                () -> (letter, cause) -> Decisions.enqueue(ThrowableCause.truncated(cause))
+            )
     );
     @SuppressWarnings("unchecked")
     private final Component<StreamableMessageSource<TrackedEventMessage<?>>> defaultStreamableSource =

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java
@@ -107,8 +107,7 @@ public class DeadLetteringEventHandlerInvoker
     public void handle(@Nonnull EventMessage<?> message, @Nonnull Segment segment) throws Exception {
         if (!super.sequencingPolicyMatchesSegment(message, segment)) {
             logger.trace("Ignoring event with id [{}] as it is not assigned to segment [{}].",
-                         message.getIdentifier(),
-                         segment);
+                         message.getIdentifier(), segment);
             return;
         }
 

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java
@@ -43,6 +43,7 @@ import java.util.function.Predicate;
 import javax.annotation.Nonnull;
 
 import static org.axonframework.common.BuilderUtils.assertNonNull;
+import static org.axonframework.messaging.deadletter.ThrowableCause.truncated;
 
 /**
  * Implementation of an {@link EventHandlerInvoker} utilizing a {@link SequencedDeadLetterQueue} to enqueue
@@ -89,7 +90,8 @@ public class DeadLetteringEventHandlerInvoker
     /**
      * Instantiate a builder to construct a {@link DeadLetteringEventHandlerInvoker}.
      * <p>
-     * The {@link EnqueuePolicy} defaults to returning {@link Decisions#enqueue(Throwable)} when invoked, the
+     * The {@link EnqueuePolicy} defaults to returning {@link Decisions#enqueue(Throwable)} that truncates the
+     * {@link Throwable#getMessage()} size to {@code 1024} characters when invoked for any dead letter, the
      * {@link ListenerInvocationErrorHandler} is defaulted to a {@link PropagatingErrorHandler}, the
      * {@link SequencingPolicy} to a {@link SequentialPerAggregatePolicy}, and {@code allowReset} defaults to
      * {@code false}. Providing at least one Event Handler, a {@link SequencedDeadLetterQueue}, and a
@@ -104,14 +106,17 @@ public class DeadLetteringEventHandlerInvoker
     @Override
     public void handle(@Nonnull EventMessage<?> message, @Nonnull Segment segment) throws Exception {
         if (!super.sequencingPolicyMatchesSegment(message, segment)) {
-            logger.trace("Ignoring event with id [{}] as it is not assigned to segment [{}].", message.getIdentifier(), segment);
+            logger.trace("Ignoring event with id [{}] as it is not assigned to segment [{}].",
+                         message.getIdentifier(),
+                         segment);
             return;
         }
 
         Object sequenceIdentifier = super.sequenceIdentifier(message);
         if (queue.enqueueIfPresent(sequenceIdentifier, () -> new GenericDeadLetter<>(sequenceIdentifier, message))) {
             if (logger.isInfoEnabled()) {
-                logger.info("Event with id [{}] is added to the dead-letter queue since its queue id [{}] is already present.",
+                logger.info("Event with id [{}] is added to the dead-letter queue "
+                                    + "since its queue id [{}] is already present.",
                             message.getIdentifier(), sequenceIdentifier);
             }
         } else {
@@ -164,7 +169,8 @@ public class DeadLetteringEventHandlerInvoker
     /**
      * Builder class to instantiate a {@link DeadLetteringEventHandlerInvoker}.
      * <p>
-     * The {@link EnqueuePolicy} defaults to returning {@link Decisions#enqueue(Throwable)} when invoked, the
+     * The {@link EnqueuePolicy} defaults to returning {@link Decisions#enqueue(Throwable)} that truncates the
+     * {@link Throwable#getMessage()} size to {@code 1024} characters when invoked for any dead letter, the
      * {@link ListenerInvocationErrorHandler} is defaulted to a {@link PropagatingErrorHandler}, the
      * {@link SequencingPolicy} to a {@link SequentialPerAggregatePolicy}, and {@code allowReset} defaults to
      * {@code false}. Providing at least one Event Handler, a {@link SequencedDeadLetterQueue}, and a
@@ -173,7 +179,7 @@ public class DeadLetteringEventHandlerInvoker
     public static class Builder extends SimpleEventHandlerInvoker.Builder<Builder> {
 
         private SequencedDeadLetterQueue<EventMessage<?>> queue;
-        private EnqueuePolicy<EventMessage<?>> enqueuePolicy = (letter, cause) -> Decisions.enqueue(cause);
+        private EnqueuePolicy<EventMessage<?>> enqueuePolicy = (letter, cause) -> Decisions.enqueue(truncated(cause));
         private TransactionManager transactionManager;
         private boolean allowReset = false;
 
@@ -199,7 +205,8 @@ public class DeadLetteringEventHandlerInvoker
         /**
          * Sets the {@link EnqueuePolicy} this {@link EventHandlerInvoker} uses to decide whether a
          * {@link DeadLetter dead letter} should be added to the {@link SequencedDeadLetterQueue}. Defaults to returning
-         * {@link Decisions#enqueue(Throwable)} when invoked for any dead letter.
+         * {@link Decisions#enqueue(Throwable)} that truncates the {@link Throwable#getMessage()} size to {@code 1024}
+         * characters when invoked for any dead letter.
          *
          * @param enqueuePolicy The {@link EnqueuePolicy} this {@link EventHandlerInvoker} uses to decide whether a
          *                      {@link DeadLetter dead letter} should be added to the {@link SequencedDeadLetterQueue}.

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/DeadLetterEntry.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/DeadLetterEntry.java
@@ -81,6 +81,7 @@ public class DeadLetterEntry {
     private Instant processingStarted;
 
     private String causeType;
+    @Column(length = 1023)
     private String causeMessage;
 
     @Basic

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/DeadLetterEntry.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/DeadLetterEntry.java
@@ -82,6 +82,7 @@ public class DeadLetterEntry {
 
     private String causeType;
     @Column(length = 1023)
+    @javax.persistence.Column(length = 1023)
     private String causeMessage;
 
     @Basic
@@ -91,7 +92,6 @@ public class DeadLetterEntry {
     @javax.persistence.Lob
     @javax.persistence.Column(length = 10000)
     private byte[] diagnostics;
-
 
     /**
      * Constructor required by JPA. Do not use.

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/JpaDeadLetter.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/JpaDeadLetter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2022. Axon Framework
+ * Copyright (c) 2010-2023. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -174,7 +174,7 @@ public class JpaDeadLetter<M extends EventMessage<?>> implements DeadLetter<M> {
                                    sequenceIdentifier,
                                    enqueuedAt,
                                    GenericDeadLetter.clock.instant(),
-                                   requeueCause != null ? new ThrowableCause(requeueCause) : cause,
+                                   requeueCause != null ? ThrowableCause.asCause(requeueCause) : cause,
                                    diagnostics,
                                    message);
     }

--- a/messaging/src/main/java/org/axonframework/eventhandling/deadletter/legacyjpa/JpaDeadLetter.java
+++ b/messaging/src/main/java/org/axonframework/eventhandling/deadletter/legacyjpa/JpaDeadLetter.java
@@ -178,7 +178,7 @@ public class JpaDeadLetter<M extends EventMessage<?>> implements DeadLetter<M> {
                                    sequenceIdentifier,
                                    enqueuedAt,
                                    GenericDeadLetter.clock.instant(),
-                                   requeueCause != null ? new ThrowableCause(requeueCause) : cause,
+                                   requeueCause != null ? ThrowableCause.asCause(requeueCause) : cause,
                                    diagnostics,
                                    message);
     }

--- a/messaging/src/main/java/org/axonframework/messaging/deadletter/GenericDeadLetter.java
+++ b/messaging/src/main/java/org/axonframework/messaging/deadletter/GenericDeadLetter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2022. Axon Framework
+ * Copyright (c) 2010-2023. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -110,7 +110,7 @@ public class GenericDeadLetter<M extends Message<?>> implements DeadLetter<M> {
     private GenericDeadLetter(GenericDeadLetter<M> delegate, Throwable requeueCause) {
         this(delegate.sequenceIdentifier,
              delegate.message(),
-             requeueCause != null ? new ThrowableCause(requeueCause) : delegate.cause,
+             requeueCause != null ? ThrowableCause.asCause(requeueCause) : delegate.cause,
              delegate.enqueuedAt(),
              clock.instant(),
              delegate.diagnostics());

--- a/messaging/src/main/java/org/axonframework/messaging/deadletter/ThrowableCause.java
+++ b/messaging/src/main/java/org/axonframework/messaging/deadletter/ThrowableCause.java
@@ -102,7 +102,10 @@ public class ThrowableCause extends AxonException implements Cause {
      * {@code messageSize}.
      */
     public static ThrowableCause truncated(Throwable throwable, int messageSize) {
-        return new ThrowableCause(throwable.getClass().getName(), throwable.getMessage().substring(0, messageSize));
+        return new ThrowableCause(
+                throwable.getClass().getName(),
+                throwable.getMessage().substring(0, Math.min(throwable.getMessage().length(), messageSize))
+        );
     }
 
     @Override

--- a/messaging/src/main/java/org/axonframework/messaging/deadletter/ThrowableCause.java
+++ b/messaging/src/main/java/org/axonframework/messaging/deadletter/ThrowableCause.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2022. Axon Framework
+ * Copyright (c) 2010-2023. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -16,6 +16,8 @@
 
 package org.axonframework.messaging.deadletter;
 
+import org.axonframework.common.AxonException;
+
 import java.util.Objects;
 
 /**
@@ -25,7 +27,13 @@ import java.util.Objects;
  * @author Mitchel Herrijgers
  * @since 4.6.0
  */
-public class ThrowableCause implements Cause {
+public class ThrowableCause extends AxonException implements Cause {
+
+    /**
+     * The default size of {@code 1023} to truncate a {@link Throwable#getMessage()} to, to fit into typical dead-letter
+     * storages.
+     */
+    public static final int TRUNCATED_MESSAGE_SIZE = 1023;
 
     private final String type;
     private final String message;
@@ -37,6 +45,7 @@ public class ThrowableCause implements Cause {
      * @param throwable The throwable to base this cause on.
      */
     public ThrowableCause(Throwable throwable) {
+        super(throwable.getMessage(), throwable);
         this.type = throwable.getClass().getName();
         this.message = throwable.getMessage();
     }
@@ -48,8 +57,52 @@ public class ThrowableCause implements Cause {
      * @param message The message of this cause.
      */
     public ThrowableCause(String type, String message) {
+        super(message);
         this.type = type;
         this.message = message;
+    }
+
+    /**
+     * Return the given {@code cause} as a {@link ThrowableCause}.
+     * <p>
+     * If the given {@code cause} is an instance of {@link ThrowableCause} it is returned as is. Otherwise, this method
+     * constructs a new instance through {@link #ThrowableCause(Throwable)}.
+     *
+     * @param cause The {@link Throwable} to map to a {@link ThrowableCause}.
+     * @return A {@link ThrowableCause} based on the given {@code cause}, or the {@code cause} as-is if it is an
+     * instance of {@code ThrowableCause}.
+     */
+    public static ThrowableCause asCause(Throwable cause) {
+        return cause instanceof ThrowableCause ? (ThrowableCause) cause : new ThrowableCause(cause);
+    }
+
+    /**
+     * Construct a {@link ThrowableCause} based on the given {@code throwable}, truncating the message to a maximum size
+     * of {@link #TRUNCATED_MESSAGE_SIZE}.
+     * <p>
+     * Should be used to ensure the {@link Cause} fits in the desired dead-letter storage solution.
+     *
+     * @param throwable The {@link Throwable} to adjust to a {@link ThrowableCause}.
+     * @return A {@link ThrowableCause} based on the given {@code throwable} for which the message is truncated to
+     * {@link #TRUNCATED_MESSAGE_SIZE}.
+     */
+    public static ThrowableCause truncated(Throwable throwable) {
+        return truncated(throwable, TRUNCATED_MESSAGE_SIZE);
+    }
+
+    /**
+     * Construct a {@link ThrowableCause} based on the given {@code throwable}, truncating the message to the given
+     * {@code messageSize}.
+     * <p>
+     * Should be used to ensure the {@link Cause} fits in the desired dead-letter storage solution.
+     *
+     * @param throwable   The {@link Throwable} to adjust to a {@link ThrowableCause}.
+     * @param messageSize The size to truncate the {@link Throwable#getMessage()} to, to be able to fit in databases.
+     * @return A {@link ThrowableCause} based on the given {@code throwable} for which the message is truncated to given
+     * {@code messageSize}.
+     */
+    public static ThrowableCause truncated(Throwable throwable, int messageSize) {
+        return new ThrowableCause(throwable.getClass().getName(), throwable.getMessage().substring(0, messageSize));
     }
 
     @Override

--- a/messaging/src/test/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventIntegrationTest.java
+++ b/messaging/src/test/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventIntegrationTest.java
@@ -28,10 +28,12 @@ import org.axonframework.eventhandling.pooled.PooledStreamingEventProcessor;
 import org.axonframework.eventhandling.tokenstore.inmemory.InMemoryTokenStore;
 import org.axonframework.messaging.MetaData;
 import org.axonframework.messaging.annotation.MessageIdentifier;
+import org.axonframework.messaging.deadletter.Cause;
 import org.axonframework.messaging.deadletter.DeadLetter;
 import org.axonframework.messaging.deadletter.Decisions;
 import org.axonframework.messaging.deadletter.EnqueuePolicy;
 import org.axonframework.messaging.deadletter.SequencedDeadLetterQueue;
+import org.axonframework.messaging.deadletter.ThrowableCause;
 import org.axonframework.messaging.unitofwork.RollbackConfigurationType;
 import org.axonframework.utils.InMemoryStreamableEventSource;
 import org.junit.jupiter.api.*;
@@ -41,6 +43,7 @@ import java.util.HashSet;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.OptionalLong;
 import java.util.Set;
 import java.util.UUID;
@@ -88,16 +91,40 @@ public abstract class DeadLetteringEventIntegrationTest {
     private static final boolean SUCCEED_RETRY = true;
     private static final boolean FAIL = false;
     private static final boolean FAIL_RETRY = false;
+    private static final String BLOB_OF_TEXT = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, "
+            + "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. "
+            + "Gravida quis blandit turpis cursus in. "
+            + "Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque fermentum. "
+            + "Egestas maecenas pharetra convallis posuere morbi leo urna. "
+            + "Dictumst quisque sagittis purus sit amet volutpat consequat. "
+            + "At volutpat diam ut venenatis tellus in metus vulputate eu. "
+            + "Imperdiet dui accumsan sit amet nulla facilisi. Eget est lorem ipsum dolor sit amet. "
+            + "Vestibulum morbi blandit cursus risus at ultrices mi tempus imperdiet. "
+            + "Sed tempus urna et pharetra pharetra massa massa. Dolor magna eget est lorem. "
+            + "Purus semper eget duis at tellus. "
+            + "Tincidunt augue interdum velit euismod in pellentesque massa placerat duis."
+            + "\n\n"
+            + "Quis ipsum suspendisse ultrices gravida dictum fusce ut. "
+            + "Nascetur ridiculus mus mauris vitae ultricies leo integer malesuada. "
+            + "Sit amet purus gravida quis blandit turpis cursus in. "
+            + "Gravida rutrum quisque non tellus. "
+            + "Eros donec ac odio tempor orci dapibus. "
+            + "Dictum varius duis at consectetur lorem donec massa sapien."
+            + "Tincidunt arcu non sodales neque sodales ut etiam sit amet. "
+            + "Sagittis aliquam malesuada bibendum arcu vitae. "
+            + "Vel turpis nunc eget lorem dolor sed viverra. "
+            + "In egestas erat imperdiet sed euismod nisi. "
+            + "Lorem ipsum dolor sit amet consectetur.";
 
     private ProblematicEventHandlingComponent eventHandlingComponent;
     private SequencedDeadLetterQueue<EventMessage<?>> deadLetterQueue;
     private DeadLetteringEventHandlerInvoker deadLetteringInvoker;
     private InMemoryStreamableEventSource eventSource;
     private StreamingEventProcessor streamingProcessor;
-    private AtomicBoolean returnReferenceErrorFromPolicy = new AtomicBoolean(false);
     protected TransactionManager transactionManager;
 
     private ScheduledExecutorService executor;
+    private final AtomicBoolean returnReferenceErrorFromPolicy = new AtomicBoolean(false);
 
     /**
      * Constructs the {@link SequencedDeadLetterQueue} implementation used during the integration test.
@@ -125,7 +152,7 @@ public abstract class DeadLetteringEventIntegrationTest {
                     decisionThrowable = new ReferenceException(UUID.randomUUID());
                 }
                 return Decisions.enqueue(
-                        decisionThrowable,
+                        ThrowableCause.truncated(decisionThrowable),
                         l -> MetaData.with("retries", (int) l.diagnostics().getOrDefault("retries", 0) + 1)
                 );
             } else {
@@ -582,6 +609,30 @@ public abstract class DeadLetteringEventIntegrationTest {
         assertEquals(ReferenceException.class.getName(), causeType);
     }
 
+    @Test
+    void largeThrowableMessageIsTruncatedUponCauseCreation() {
+        String aggregateId = UUID.randomUUID().toString();
+        String truncatedText = "truncated-text";
+        String testCauseMessage = BLOB_OF_TEXT + truncatedText;
+
+        eventSource.publishMessage(
+                asEventMessage(new DeadLetterableEvent(aggregateId, FAIL, FAIL, testCauseMessage))
+        );
+
+        startProcessingEvent();
+        await().pollDelay(Duration.ofMillis(25))
+               .atMost(Duration.ofSeconds(1))
+               .until(() -> deadLetterQueue.amountOfSequences() == 1);
+
+        DeadLetter<?> deadLetter = deadLetterQueue.deadLetters().iterator().next().iterator().next();
+        Optional<Cause> optionalCause = deadLetter.cause();
+        assertTrue(optionalCause.isPresent());
+        String resultMessage = optionalCause.get().message();
+        assertNotEquals(testCauseMessage, resultMessage);
+        assertFalse(resultMessage.contains(truncatedText));
+        assertTrue(resultMessage.contains(BLOB_OF_TEXT.substring(0, 10)));
+    }
+
     private void publishEventsFor(String aggregateId,
                                   int immediateSuccessesPerAggregate,
                                   int failFirstAndThenSucceedPerAggregate,
@@ -643,7 +694,10 @@ public abstract class DeadLetteringEventIntegrationTest {
                 firstTrySuccesses.compute(sequenceId, (id, count) -> count == null ? 1 : ++count);
             } else {
                 firstTryFailures.compute(sequenceId, (id, count) -> count == null ? 1 : ++count);
-                throw new RuntimeException("Initial handling failed. Let's dead letter event [" + sequenceId + "].");
+                throw new RuntimeException(
+                        "Initial handling failed. Let's dead letter event [" + sequenceId + "].\n"
+                                + event.getCauseMessage()
+                );
             }
         }
 
@@ -652,7 +706,9 @@ public abstract class DeadLetteringEventIntegrationTest {
                 evaluationSuccesses.compute(sequenceId, (id, count) -> count == null ? 1 : ++count);
             } else {
                 evaluationFailures.compute(sequenceId, (id, count) -> count == null ? 1 : ++count);
-                throw new RuntimeException("Evaluation failed. Let's dead letter event [" + sequenceId + "].");
+                throw new RuntimeException(
+                        "Evaluation failed. Let's dead letter event [" + sequenceId + "].\n" + event.getCauseMessage()
+                );
             }
         }
 
@@ -702,19 +758,28 @@ public abstract class DeadLetteringEventIntegrationTest {
         private final String aggregateIdentifier;
         private final boolean shouldSucceed;
         private final boolean shouldSucceedOnEvaluation;
+        private final String causeMessage;
 
         private DeadLetterableEvent(String aggregateIdentifier,
                                     boolean shouldSucceed) {
             this(aggregateIdentifier, shouldSucceed, true);
         }
 
+        private DeadLetterableEvent(String aggregateIdentifier,
+                                    boolean shouldSucceed,
+                                    boolean shouldSucceedOnEvaluation) {
+            this(aggregateIdentifier, shouldSucceed, shouldSucceedOnEvaluation, "");
+        }
+
         @JsonCreator
         public DeadLetterableEvent(@JsonProperty("aggregateIdentifier") String aggregateIdentifier,
                                    @JsonProperty("shouldSucceed") boolean shouldSucceed,
-                                   @JsonProperty("shouldSucceedOnEvaluation") boolean shouldSucceedOnEvaluation) {
+                                   @JsonProperty("shouldSucceedOnEvaluation") boolean shouldSucceedOnEvaluation,
+                                   @JsonProperty("causeMessage") String causeMessage) {
             this.aggregateIdentifier = aggregateIdentifier;
             this.shouldSucceed = shouldSucceed;
             this.shouldSucceedOnEvaluation = shouldSucceedOnEvaluation;
+            this.causeMessage = causeMessage;
         }
 
         public String getAggregateIdentifier() {
@@ -729,6 +794,10 @@ public abstract class DeadLetteringEventIntegrationTest {
             return shouldSucceedOnEvaluation;
         }
 
+        public String getCauseMessage() {
+            return causeMessage;
+        }
+
         @Override
         public boolean equals(Object o) {
             if (this == o) {
@@ -738,14 +807,15 @@ public abstract class DeadLetteringEventIntegrationTest {
                 return false;
             }
             DeadLetterableEvent that = (DeadLetterableEvent) o;
-            return shouldSucceed == that.shouldSucceed
-                    && shouldSucceedOnEvaluation == that.shouldSucceedOnEvaluation
-                    && Objects.equals(aggregateIdentifier, that.aggregateIdentifier);
+            return shouldSucceed == that.shouldSucceed && shouldSucceedOnEvaluation == that.shouldSucceedOnEvaluation
+                    && Objects.equals(aggregateIdentifier, that.aggregateIdentifier) && Objects.equals(
+                    causeMessage,
+                    that.causeMessage);
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(aggregateIdentifier, shouldSucceed, shouldSucceedOnEvaluation);
+            return Objects.hash(aggregateIdentifier, shouldSucceed, shouldSucceedOnEvaluation, causeMessage);
         }
 
         @Override
@@ -754,6 +824,7 @@ public abstract class DeadLetteringEventIntegrationTest {
                     "aggregateIdentifier='" + aggregateIdentifier + '\'' +
                     ", shouldSucceed=" + shouldSucceed +
                     ", shouldSucceedOnEvaluation=" + shouldSucceedOnEvaluation +
+                    ", causeMessage='" + causeMessage + '\'' +
                     '}';
         }
     }

--- a/messaging/src/test/java/org/axonframework/messaging/deadletter/ThrowableCauseTest.java
+++ b/messaging/src/test/java/org/axonframework/messaging/deadletter/ThrowableCauseTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2022. Axon Framework
+ * Copyright (c) 2010-2023. Axon Framework
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -27,6 +27,31 @@ import static org.junit.jupiter.api.Assertions.*;
  */
 class ThrowableCauseTest {
 
+    private static final String BLOB_OF_TEXT = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, "
+            + "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. "
+            + "Gravida quis blandit turpis cursus in. "
+            + "Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque fermentum. "
+            + "Egestas maecenas pharetra convallis posuere morbi leo urna. "
+            + "Dictumst quisque sagittis purus sit amet volutpat consequat. "
+            + "At volutpat diam ut venenatis tellus in metus vulputate eu. "
+            + "Imperdiet dui accumsan sit amet nulla facilisi. Eget est lorem ipsum dolor sit amet. "
+            + "Vestibulum morbi blandit cursus risus at ultrices mi tempus imperdiet. "
+            + "Sed tempus urna et pharetra pharetra massa massa. Dolor magna eget est lorem. "
+            + "Purus semper eget duis at tellus. "
+            + "Tincidunt augue interdum velit euismod in pellentesque massa placerat duis."
+            + "\n\n"
+            + "Quis ipsum suspendisse ultrices gravida dictum fusce ut. "
+            + "Nascetur ridiculus mus mauris vitae ultricies leo integer malesuada. "
+            + "Sit amet purus gravida quis blandit turpis cursus in. "
+            + "Gravida rutrum quisque non tellus. "
+            + "Eros donec ac odio tempor orci dapibus. "
+            + "Dictum varius duis at consectetur lorem donec massa sapien."
+            + "Tincidunt arcu non sodales neque sodales ut etiam sit amet. "
+            + "Sagittis aliquam malesuada bibendum arcu vitae. "
+            + "Vel turpis nunc eget lorem dolor sed viverra. "
+            + "In egestas erat imperdiet sed euismod nisi. "
+            + "Lorem ipsum dolor sit amet consectetur.";
+
     @Test
     void constructThrowableCauseWithThrowable() {
         Throwable testThrowable = new RuntimeException("some-message");
@@ -46,5 +71,71 @@ class ThrowableCauseTest {
 
         assertEquals(testType, testSubject.type());
         assertEquals(testMessage, testSubject.message());
+    }
+
+    @Test
+    void asCauseWrapsThrowable() {
+        Throwable testThrowable = new RuntimeException("some-message");
+
+        ThrowableCause result = ThrowableCause.asCause(testThrowable);
+
+        assertEquals(testThrowable.getClass().getName(), result.type());
+        assertEquals(testThrowable.getMessage(), result.message());
+    }
+
+    @Test
+    void asCauseReturnsCauseAsIs() {
+        ThrowableCause testCause = new ThrowableCause("type", "message");
+
+        ThrowableCause result = ThrowableCause.asCause(testCause);
+
+        assertEquals(testCause, result);
+    }
+
+    @Test
+    void truncateLeavesMessageAsIsIfItDoesNotExceedMessageSize() {
+        Throwable testThrowable = new RuntimeException("some-message");
+
+        ThrowableCause result = ThrowableCause.truncated(testThrowable);
+
+        assertEquals(testThrowable.getClass().getName(), result.type());
+        assertEquals(testThrowable.getMessage(), result.message());
+    }
+
+    @Test
+    void truncateShortensMessageIfItExceedsMessageSize() {
+        String textThatShouldBeTruncated = "truncated-text-at-the-end";
+        Throwable testThrowable = new RuntimeException(BLOB_OF_TEXT + textThatShouldBeTruncated);
+
+        ThrowableCause result = ThrowableCause.truncated(testThrowable);
+
+        assertEquals(testThrowable.getClass().getName(), result.type());
+        assertNotEquals(testThrowable.getMessage(), result.getMessage());
+        assertFalse(result.getMessage().contains(textThatShouldBeTruncated));
+    }
+
+    @Test
+    void truncateWithSpecifiedMessageSizeLeavesMessageAsIsIfItDoesNotExceedGivenSize() {
+        String testMessage = "some-message";
+        Throwable testThrowable = new RuntimeException(testMessage);
+
+        ThrowableCause result = ThrowableCause.truncated(testThrowable, testMessage.length());
+
+        assertEquals(testThrowable.getClass().getName(), result.type());
+        assertEquals(testThrowable.getMessage(), result.message());
+    }
+
+    @Test
+    void truncateWithSpecifiedMessageSizeShortensMessageIfItExceedsGivenSize() {
+        String textThatShouldBeTruncated = "truncated-text-at-the-end";
+        Throwable testThrowable = new RuntimeException(textThatShouldBeTruncated);
+        int testMessageSize = textThatShouldBeTruncated.length() - 3;
+
+        ThrowableCause result = ThrowableCause.truncated(testThrowable, testMessageSize);
+
+        assertEquals(testThrowable.getClass().getName(), result.type());
+        assertNotEquals(testThrowable.getMessage(), result.getMessage());
+        assertFalse(result.getMessage().contains(textThatShouldBeTruncated));
+        assertTrue(result.getMessage().contains(textThatShouldBeTruncated.substring(0, testMessageSize)));
     }
 }

